firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.remote;
+
+import android.util.Base64;
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class BloomFilter {
+  private final int bitCount;
+  private final byte[] bitmap;
+  private final int hashCount;
+  private final MessageDigest md5HashMessageDigest;
+
+  public BloomFilter(@NonNull byte[] bitmap, int padding, int hashCount) {
+    if (bitmap == null) {
+      throw new NullPointerException("Bitmap cannot be null.");
+    }
+    if (padding < 0 || padding >= 8) {
+      throw new BloomFilterException("Invalid padding: " + padding);
```

Comment:
I've been thinking about this and I want to change this pattern to be more like the iOS SDK, where we created a "factory" function to provide friendly error reporting. It's pretty unconventional in Android to throw a custom exception from a constructor like this. I'll open a PR for it.